File: Linked Lists/palindrome.js
import { factoryStack as createStack } from '../../Data Structures/stack.js';

/**
 * Solution 1 - Hash map
 *
 * Complexity Analysis
 * O(n) time | O(n) space
 *
 * Check if a linked list is a palindrome
 *
 * @param {object} linkedList
 * @returns {bool}
 */
function isPalindrome(linkedList = {}) {
    // Gracefully handle type and Falsy values
    if (typeof linkedList !== 'object' || Object.keys(linkedList || {}).length === 0) {
        console.error('Parameter must be a valid non-empty object');
        return false;
    }

    const charsMap = new Map();

    let currNode = linkedList;

    while (currNode !== null) {
        const { data } = currNode;
        const char = data.toLowerCase();

        currNode = currNode.next;

        if (char === ' ') {
            continue;
        }

        if (charsMap.has(char)) {
            charsMap.set(char, charsMap.get(char) + 1);
        } else {
            charsMap.set(char, 1);
        }
    }

    let oddCount = 0;

    for (const char of charsMap.values()) {
        if (char % 2 === 1) {
            oddCount += 1;
        }
    }

    return oddCount <= 1;
}

/**
 * Solution 2 - Reverse string
 *
 * Complexity Analysis
 * O(n) time | O(n) space
 *
 * Check if a linked list is a palindrome
 *
 * @param {object} linkedList
 * @returns {bool}
 */
function isPalindrome(linkedList = {}) {
    // Gracefully handle type and Falsy values
    if (typeof linkedList !== 'object' || Object.keys(linkedList || {}).length === 0) {
        console.error('Parameter must be a valid non-empty object');
        return false;
    }

    let slowPointer = linkedList;
    let fastPointer = linkedList;

    let middleNode = null;

    let stringFirstHalf = '';
    let stringSecondHalf = '';

    while (slowPointer !== null) {
        if (fastPointer !== null && fastPointer.next !== null) {
            stringFirstHalf = slowPointer.data + stringFirstHalf;

            middleNode = slowPointer.next;
            fastPointer = fastPointer.next.next;
        } else {
            stringSecondHalf += slowPointer.data;
        }

        slowPointer = slowPointer.next;
    }

    // Fast pointer will NOT be NULL if there is odd elements in the linked list
    if (fastPointer !== null) {
        stringFirstHalf = middleNode.data + stringFirstHalf;
    }

    return stringFirstHalf.toLowerCase() === stringSecondHalf.toLowerCase();
}

/**
 * Solution 3 - Runner slow/fast pointers creating a stack
 *
 * Complexity Analysis
 * O(n) time | O(n) space
 *
 * Check if a linked list is a palindrome
 *
 * @param {object} linkedList
 * @returns {bool}
 */
function isPalindrome(linkedList = {}) {
    // Gracefully handle type and Falsy values
    if (typeof linkedList !== 'object' || Object.keys(linkedList || {}).length === 0) {
        console.error('Parameter must be a valid non-empty object');
        return false;
    }

    const charsStack = createStack();

    let slowPointer = linkedList;
    let fastPointer = linkedList;

    let middleNode = null;

    while (slowPointer !== null) {
        if (fastPointer !== null && fastPointer.next !== null) {
            charsStack.push(slowPointer.data);

            middleNode = slowPointer.next;
            fastPointer = fastPointer.next.next;
        } else {
            // Skip middle node in odd linked list
            if (fastPointer !== null && (middleNode === slowPointer)) {
                slowPointer = slowPointer.next;
                continue;
            }

            const charStack = charsStack.pop();

            if (slowPointer.data.toLowerCase() !== charStack.value.toLowerCase()) {
                return false;
            }
        }

        slowPointer = slowPointer.next;
    }

    return true;
}

const testCases = [
    // Normal
    // Data that is typical (expected) and should be accepted by the system.
    {
        assert: isPalindrome({ data: 'a', next: { data: 'b', next: { data: 'b', next: { data: 'a', next: null } } } }),
        expected: true,
    },
    {
        assert: isPalindrome({ data: 'a', next: { data: 'b', next: { data: 'c', next: { data: 'b', next: { data: 'a', next: null } } } } }),
        expected: true,
    },
    {
        assert: isPalindrome({ data: 'a', next: { data: 'b', next: { data: 'c', next: { data: 'd', next: { data: 'e', next: null } } } } }),
        expected: false
    },
    {
        assert: isPalindrome({ data: 'K', next: { data: 'a', next: { data: 'y', next: { data: 'a', next: { data: 'k', next: null } } } } }),
        expected: true
    },
    {
        assert: isPalindrome({ data: 'R', next: { data: 'e', next: { data: 'p', next: { data: 'a', next: { data: 'p', next: { data: 'e', next: { data: 'r', next: null } } } } } } }),
        expected: true
    },
    {
        assert: isPalindrome({ data: 'T', next: { data: 'a', next: { data: 'c', next: { data: 'o', next: { data: 'C', next: { data: 'a', next: { data: 't', next: null } } } } } } }),
        expected: true
    },
    {
        assert: isPalindrome({ data: 'T', next: { data: 'a', next: { data: 'c', next: { data: 'o', next: { data: 'C', next: { data: 'a', next: { data: 't', next: { data: 'w', next: null } } } } } } } }),
        expected: false
    },

    // Boundary data (extreme data, edge case)
    // Data at the upper or lower limits of expectations that should be accepted by the system.
    {
        assert: isPalindrome({ data: 'a', next: null }),
        expected: true
    },

    // Abnormal data (erroneous data)
    // Data that falls outside of what is acceptable and should be rejected by the system.
    { assert: isPalindrome(), expected: false },
    { assert: isPalindrome(0), expected: false },
    { assert: isPalindrome(''), expected: false },
    { assert: isPalindrome([]), expected: false },
    { assert: isPalindrome({}), expected: false },
    { assert: isPalindrome(NaN), expected: false },
    { assert: isPalindrome(null), expected: false },
    { assert: isPalindrome(false), expected: false },
    { assert: isPalindrome(new Set()), expected: false },
    { assert: isPalindrome(new Map()), expected: false },
    { assert: isPalindrome(undefined), expected: false },
];

// Run tests
testCases.forEach((test, index) => {
    const currentTest = `# Test ${index + 1}`;

    console.group(currentTest);
    console.log('Actual:', test.assert);
    console.log('Expected:', test.expected);
    console.log(test.assert === test.expected ? '🤘 Test PASSED 🤘' : '👎 Test FAILED 👎', '\n');
    console.groupEnd(currentTest);
});

File: Linked Lists/palindrome.py
import sys
sys.path.append('../../Data Structures')

from stack import Stack

def isPalindrome(linkedList={}):
    '''
    Solution 1 - Hash map

    Complexity Analysis
    O(n) time | O(n) space

    Check if a linked list is a palindrome

    dict: linkedList
    return: True if its palindrome
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(linkedList, dict) or bool(linkedList) == False):
        print('Argument should be a valid non-empty dictionary')
        return False

    charsMap = dict()
    currNode = linkedList

    while (currNode != None):
        char = currNode['data']
        charLower = char.lower()

        currNode = currNode['next']

        if (char == ' '):
            continue

        if (charLower in charsMap):
            charsMap[charLower] += 1
        else:
            charsMap[charLower] = 1

    oddCount = 0

    for char in charsMap.values():
        if (char % 2 == 1):
            oddCount += 1

    return oddCount <= 1

def isPalindrome(linkedList={}):
    '''
    Solution 2 - Reverse string

    Complexity Analysis
    O(n) time | O(n) space

    Check if a linked list is a palindrome

    dict: linkedList
    return: True if its palindrome
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(linkedList, dict) or bool(linkedList) == False):
        print('Argument should be a valid non-empty dictionary')
        return False

    slowPointer = linkedList
    fastPointer = linkedList

    middleNode = None

    stringFirstHalf = ''
    stringSecondHalf = ''

    while (slowPointer != None):
        if (fastPointer != None and fastPointer['next'] != None):
            stringFirstHalf = slowPointer['data'] + stringFirstHalf

            middleNode = slowPointer['next']
            fastPointer = fastPointer['next']['next']
        else:
            stringSecondHalf += slowPointer['data']

        slowPointer = slowPointer['next']

    # Fast pointer will NOT be NULL if there is odd elements in the linked list
    if (fastPointer != None):
        stringFirstHalf = middleNode['data'] + stringFirstHalf

    return stringFirstHalf.lower() == stringSecondHalf.lower()

def isPalindrome(linkedList={}):
    '''
    Solution 3 - Runner slow/fast pointers creating a stack

    Complexity Analysis
    O(n) time | O(n) space

    Check if a linked list is a palindrome

    dict: linkedList
    return: True if its palindrome
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(linkedList, dict) or bool(linkedList) == False):
        print('Argument should be a valid non-empty dictionary')
        return False

    charsStack = Stack()

    slowPointer = linkedList
    fastPointer = linkedList

    middleNode = None

    while (slowPointer != None):
        if (fastPointer != None and fastPointer['next'] != None):
            charsStack.push(slowPointer['data'])

            middleNode = slowPointer['next']
            fastPointer = fastPointer['next']['next']
        else:
            # Skip middle node in odd linked list
            if (fastPointer != None and middleNode == slowPointer):
                slowPointer = slowPointer['next']
                continue

            charStack = charsStack.pop()

            if (slowPointer['data'].lower() != charStack.lower()):
                return False

        slowPointer = slowPointer['next']

    return True


# Test cases (black box - unit testing)
testCases = [
    # Normal
    # Data that is typical (expected) and should be accepted by the system.
    {
        'assert': isPalindrome({ 'data': 'a', 'next': { 'data': 'b', 'next': { 'data': 'b', 'next': { 'data': 'a', 'next': None } } } }),
        'expected': True,
    },
    {
        'assert': isPalindrome({ 'data': 'a', 'next': { 'data': 'b', 'next': { 'data': 'c', 'next': { 'data': 'b', 'next': { 'data': 'a', 'next': None } } } } }),
        'expected': True,
    },
    {
        'assert': isPalindrome({ 'data': 'a', 'next': { 'data': 'b', 'next': { 'data': 'c', 'next': { 'data': 'd', 'next': { 'data': 'e', 'next': None } } } } }),
        'expected': False
    },
    {
        'assert': isPalindrome({ 'data': 'K', 'next': { 'data': 'a', 'next': { 'data': 'y', 'next': { 'data': 'a', 'next': { 'data': 'k', 'next': None } } } } }),
        'expected': True
    },
    {
        'assert': isPalindrome({ 'data': 'R', 'next': { 'data': 'e', 'next': { 'data': 'p', 'next': { 'data': 'a', 'next': { 'data': 'p', 'next': { 'data': 'e', 'next': { 'data': 'r', 'next': None } } } } } } }),
        'expected': True
    },
    {
        'assert': isPalindrome({ 'data': 'T', 'next': { 'data': 'a', 'next': { 'data': 'c', 'next': { 'data': 'o', 'next': { 'data': 'C', 'next': { 'data': 'a', 'next': { 'data': 't', 'next': None } } } } } } }),
        'expected': True
    },
    {
        'assert': isPalindrome({ 'data': 'T', 'next': { 'data': 'a', 'next': { 'data': 'c', 'next': { 'data': 'o', 'next': { 'data': 'C', 'next': { 'data': 'a', 'next': { 'data': 't', 'next': { 'data': 'w', 'next': None } } } } } } } }),
        'expected': False
    },
    # Boundary data (extreme data, edge case)
    # Data at the upper or lower limits of expectations that should be accepted by the system.
    # {
    #     'assert': isPalindrome({ 'data': 'a', 'next': None }),
    #     'expected': True
    # },

    # Abnormal data (erroneous data)
    # Data that falls outside of what is acceptable and should be rejected by the system.
    { 'assert': isPalindrome(), 'expected': False },
    { 'assert': isPalindrome(0), 'expected': False },
    { 'assert': isPalindrome(''), 'expected': False },
    { 'assert': isPalindrome([]), 'expected': False },
    { 'assert': isPalindrome(()), 'expected': False },
    { 'assert': isPalindrome({}), 'expected': False },
    { 'assert': isPalindrome(None), 'expected': False },
    { 'assert': isPalindrome(False), 'expected': False }
]

# Run tests
for (index, test) in enumerate(testCases):
    print(f'# Test {index + 1}')
    print(f'Actual: {test["assert"]}')
    print(f'Expected: {test["expected"]}')
    print('🤘 Test PASSED 🤘' if test['assert'] == test['expected'] else '👎 Test FAILED 👎', '\n')


